File: Editorial/3453_Separate_Squares_I/Go/2_Scanning_Line.go
func separateSquares(squares [][]int) float64 {
    var totalArea int64 = 0
    type Event struct {
        y     int
        l     int
        delta int
    }
    events := make([]Event, 0, len(squares)*2)
    
    for _, sq := range squares {
        y, l := sq[1], sq[2]
        totalArea += int64(l) * int64(l)
        events = append(events, Event{y, l, 1})
        events = append(events, Event{y + l, l, -1})
    }
    
    // sort by y-coordinate
    sort.Slice(events, func(i, j int) bool {
        return events[i].y < events[j].y
    })
    
    coveredWidth := 0.0  // sum of all bottom edges under the current scanning line
    currArea := 0.0      // current cumulative area
    prevHeight := 0.0    // height of the previous scanning line
    
    for _, event := range events {
        y, l, delta := event.y, event.l, event.delta
        diff := float64(y) - prevHeight
        // additional area between two scanning lines
        area := coveredWidth * diff
        // if this part of the area exceeds more than half of the total area
        if 2.0*(currArea+area) >= float64(totalArea) {
            return prevHeight + (float64(totalArea) - 2.0*currArea) / (2.0 * coveredWidth)
        }
        // update width: add width at the start event, subtract width at the end event
        coveredWidth += float64(delta * l)
        currArea += area
        prevHeight = float64(y)
    }
    
    return 0.0
}

File: Editorial/3401-3500/3453_Separate_Squares_I/TypeScript/2_Scanning_Line.ts
function separateSquares(squares: number[][]): number {
    let totalArea: bigint = 0n;
    const events: [number, number, number][] = [];

    for (const sq of squares) {
        const y = sq[1],
            l = sq[2];
        totalArea += BigInt(l) * BigInt(l);
        events.push([y, l, 1]);
        events.push([y + l, l, -1]);
    }

    // sort by y-coordinate
    events.sort((a, b) => a[0] - b[0]);

    let coveredWidth: number = 0; // sum of all bottom edges under the current scanning line
    let currArea: number = 0; // current cumulative area
    let prevHeight: number = 0; // height of the previous scanning line

    for (const [y, l, delta] of events) {
        const diff: number = y - prevHeight;
        // additional area between two scanning lines
        const area: number = coveredWidth * diff;
        // if this part of the area exceeds more than half of the total area
        if (2n * BigInt(Math.ceil(currArea + area)) >= totalArea) {
            return (
                prevHeight +
                (Number(totalArea) - 2.0 * currArea) / (2.0 * coveredWidth)
            );
        }
        // update width: add width at the start event, subtract width at the end event
        coveredWidth += delta * l;
        currArea += area;
        prevHeight = y;
    }

    return 0.0;
}
